codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -16,329 +16,506 @@
+                    return true;
+                }
+                try {
+                    return streamVisitor.visit(stream2);
+                } catch (Throwable cause) {
+                    onError(ctx, cause);
+                    return false;
+                }
+            }
+        });
+    }
+
     @Override
-    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
-        ctx.pipeline().remove(http2Handler);
+    public final void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        this.ctx = ctx;
+        super.handlerAdded(ctx);
+        handlerAdded0(ctx);
+        // Must be after Http2ConnectionHandler does its initialization in handlerAdded above.
+        // The server will not send a connection preface so we are good to send a window update.
+        Http2Connection connection = connection();
+        if (connection.isServer()) {
+            tryExpandConnectionFlowControlWindow(connection);
+        }
+    }
+
+    private void tryExpandConnectionFlowControlWindow(Http2Connection connection) throws Http2Exception {
+        if (initialFlowControlWindowSize != null) {
+            // The window size int the settings explicitly excludes the connection window. So we manually manipulate the
+            // connection window to accommodate more concurrent data per connection.
+            Http2Stream connectionStream = connection.connectionStream();
+            Http2LocalFlowController localFlowController = connection.local().flowController();
+            final int delta = initialFlowControlWindowSize - localFlowController.initialWindowSize(connectionStream);
+            // Only increase the connection window, don't decrease it.
+            if (delta > 0) {
+                // Double the delta just so a single stream can't exhaust the connection window.
+                localFlowController.incrementWindowSize(connectionStream, Math.max(delta << 1, delta));
+                flush(ctx);
+            }
+        }
+    }
+
+    void handlerAdded0(@SuppressWarnings("unsed") ChannelHandlerContext ctx) throws Exception {
+        // sub-class can override this for extra steps that needs to be done when the handler is added.
     }
 
     /**
      * Handles the cleartext HTTP upgrade event. If an upgrade occurred, sends a simple response via
      * HTTP/2 on stream 1 (the stream specifically reserved for cleartext HTTP upgrade).
      */
     @Override
-    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
-        if (!(evt instanceof UpgradeEvent)) {
+    public final void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (evt instanceof Http2ConnectionPrefaceWrittenEvent) {
+            // The user event implies that we are on the client.
+            tryExpandConnectionFlowControlWindow(connection());
+        } else if (evt instanceof UpgradeEvent) {
```

Comment:
any reason why we don't propagate the `Http2ConnectionPrefaceWrittenEvent` event through the pipeline? Folks may use this to know when its OK to start writing data.